### PR TITLE
Support resnet

### LINKED ISSUE #40: Implement CIFAR100 sample

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: src/runtime/cpu/layers/cpuadd.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include "runtime/cpu/layers/cpuadd.hpp"
+#include "runtime/cpu/kernels/kernels.hpp"
+#include <iostream>
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpuadd.cpp:3]: remove

File: samples/cifar100/main.cpp [hunk 2]
```diff
@@ -6,39 +6,66 @@
...
+                               std::array<int, 2>{3, 3},
+                               std::array<int, 2>{1, 1},
+                               std::array<int, 2>{stride, stride})(x);
+
+     out = dw::BatchNorm2D(0.001, 0.05)(out);
+     out = dw::ReLU()(out);
+     out = dw::Convolution(c_out,
+                           std::array<int, 2>{3, 3},
+                           std::array<int, 2>{1, 1},
+                           std::array<int, 2>{stride, stride})(out);
```
Comment (l-bat) [inline samples/cifar100/main.cpp]:
```suggestion
                           std::array<int, 2>{1, 1})(out);
```

File: samples/cifar100/main.cpp [hunk 3]
```diff
@@ -6,39 +6,66 @@
...
+                              std::array<int, 2>{1, 1})(in);
+    out = dw::BatchNorm2D(0.001, 0.05)(out);
+    out = dw::ReLU()(out);
 
-    out = dw::Linear(mid_features_first, "linear7")(out);
-    out = dw::ReLU("relu8")(out);
-    out = dw::BatchNorm1D(0.001, 0.05, "batchnorm1d9")(out);
+    out = make_layer(out, 16, 1);
+    out = make_layer(out, 32, 1);
+    out = make_layer(out, 64, 1);
```
Comment (l-bat) [inline samples/cifar100/main.cpp]:
```suggestion
    out = make_layer(out, 16, 2);
    out = make_layer(out, 32, 2);
    out = make_layer(out, 64, 2);
```

**Review by l-bat — APPROVED** [review]
Can we delete test?

File: src/runtime/cpu/layers/cpubatchnorm1d.cpp [hunk 4]
```diff
@@ -3,6 +3,8 @@
 #include <deepworks/initializers.hpp>
 #include <deepworks/utils/utils.hpp>
 
+#include <iostream>
+
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpubatchnorm1d.cpp:7]: remove

File: src/runtime/cpu/layers/cpubatchnorm2d.cpp [hunk 5]
```diff
@@ -3,6 +3,8 @@
 #include <deepworks/initializers.hpp>
 #include <deepworks/utils/utils.hpp>
 
+#include <iostream>
+
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpubatchnorm2d.cpp:7]: remove

File: src/runtime/cpu/layers/cpuconvolution.cpp [hunk 6]
```diff
@@ -1,5 +1,6 @@
 #include "runtime/cpu/layers/cpuconvolution.hpp"
 #include "runtime/cpu/kernels/kernels.hpp"
+#include <iostream>
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpuconvolution.cpp:3]: remove

File: tests/unit/kernels_reference.cpp [hunk 7]
```diff
@@ -100,10 +100,10 @@ void dw::reference::CPUReLUForward(const float* in, float* out, size_t size) {
...
 }
 
-void dw::reference::CPUReLUBackward(const float* in, const float* grad_output,
-                                    float* grad_input, size_t batch_size, size_t features) {
-    std::transform(in, in + batch_size * features, grad_output, grad_input,
-                   [](float in, float go) { return in > 0.0 ? go : 0.0;});
+void dw::reference::CPUReLUBackward(const float* out, const float* grad_output,
+                                    float* grad_input, size_t size) {
+    std::transform(out, out + size, grad_output, grad_input,
+                   [](float out, float go) { return out > 0.0 ? go : 0.0;});
```
Comment (l-bat) [inline tests/unit/kernels_reference.cpp:106]: Why not `in`?

File: tests/unit/kernels_reference.cpp [hunk 8]
```diff
@@ -628,3 +628,25 @@ void dw::reference::CPUBatchNorm2DBackward(const dw::Tensor& input_centered,
...
+    DeepWorks_Assert(input0.shape() == input1.shape() &&
+                     input0.shape() == output.shape());
+
+    const auto* in0_p = input0.data();
+    const auto* in1_p = input1.data();
+          auto* out_p = output.data();
+
+    for (int i = 0; i < input0.total(); ++i) {
+        out_p[i] = in0_p[i] + in1_p[i];
+    }
```
Comment (l-bat) [inline tests/unit/kernels_reference.cpp:644]: Can we add omp here?

File: tests/unit/test_networks.cpp [hunk 9]
```diff
@@ -769,3 +769,540 @@ TEST_F(CIFAR10Model, TrainLoopSmoke) {
...
+    void forward(const dw::Tensor& input, dw::Tensor& output) {
+        dw::reference::CPUBatchNorm2DForward(input,
+                                             output,
+                                             expected.at("gamma").data(),
+                                             expected.at("beta").data(),
+                                             moving_mean,
+                                             moving_var,
+                                             input_centered,
+                                             std,
+                                             true,
```
Comment (TolyaTalamanov) [inline tests/unit/test_networks.cpp:841]: Can be tested `eval` mode as well, in future :)

File: src/runtime/cpu/layers/cpubatchnorm1d.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpubatchnorm1d.cpp]: Done

File: src/runtime/cpu/layers/cpuconvolution.cpp (hunk 6 above)
Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpuconvolution.cpp]: Done

File: samples/cifar100/main.cpp [hunk 10]
```diff
@@ -6,39 +6,69 @@
 
 namespace dw = deepworks;
 
-static dw::Model buildCIFAR100Model(int batch_size) {
-    int mid_features_first = 384;
-    int mid_features_second = 192;
-    int out_features = 100;
+float eps = 1e-5;
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply samples/cifar100/main.cpp]: Done

File: src/io/image_reader.cpp [hunk 11]
```diff
@@ -10,10 +10,10 @@
 
 #ifdef HAVE_PNG
 #include <png.h>
-#include <algorithm>
-
 #endif
 
+#include <algorithm> // std::copy_n
```
Comment (TolyaTalamanov) [inline src/io/image_reader.cpp:15]: It must be not under compile condition

File: src/runtime/cpu/kernels/kernels.cpp [hunk 12]
```diff
@@ -138,12 +138,10 @@ void deepworks::CPUConvolutionalInputGrad(const Tensor& grad_output,
...
     int output_offset = c_out * h_out * w_out;
 
-    ConstMatrix weights_mat{weights.data(), c_out, rows};
-    Tensor grad_im2col_buf;
-    grad_im2col_buf.allocate({rows, cols});
-
-//    #pragma omp parallel for
+    #pragma omp parallel for
     for (size_t b = 0; b < batch; b++) {
+        ConstMatrix weights_mat{weights.data(), c_out, rows};
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:143]: These stuff must be thread local

**Comment by TolyaTalamanov** [comment]
@l-bat Have a look, please

File: src/runtime/cpu/kernels/kernels.cpp (hunk 12 above)
Comment (l-bat) [reply src/runtime/cpu/kernels/kernels.cpp:143]: Great!

File: src/runtime/cpu/layers/cpuadd.cpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+    #pragma omp parallel for
+    for (int i = 0; i < go.total(); ++i) {
+        gi_0.data()[i] += go.data()[i];
+    }
+
+    #pragma omp parallel for
+    for (int i = 0; i < go.total(); ++i) {
+        gi_1.data()[i] += go.data()[i];
+    }
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpuadd.cpp:29]: Can we do it in one loop?

Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpuadd.cpp]: Yep

**Pull request merged**